lib/CLI/Parser.hs

```diff
@@ -0,0 +1,550 @@
+    <*> many (option readFilter $ mconcat
+          [ metavar "FILTER"
+          , long "filter"
+          , helpDoc $ Just $ Pretty.vsep
+            [ "Filter entries."
+            , expectedFilterFormat
+            ]
+          ])
+    <*> many (option readFilterField $ mconcat
+          [ metavar "FILTERFIELD"
+          , long "filter-field"
+          , helpDoc $ Just $ Pretty.vsep
+            [ "Filter entries based on a field."
+            , expectedFilterFieldFormat
+            ]
+          ])
+
+renameOptions :: Parser RenameOptions
+renameOptions =
+  RenameOptions
+    <$> argument readPath ( mconcat
+          [ metavar "OLDPATH"
+          , help "The path to move the old directory or entry from"
+          ])
+    <*> argument readPath ( mconcat
+          [ metavar "NEWPATH"
+          , help "The path to move the directory or entry to"
+          ])
+    <*> switch ( mconcat
+          [ long "force"
+          , short 'f'
+          , help "If a directory or entry already exists under the specified path, delete it and insert this entry instead"
+          ])
+
+copyOptions :: Parser CopyOptions
+copyOptions =
+  CopyOptions
+    <$> argument readPath ( mconcat
+          [ metavar "OLDPATH"
+          , help "The path to copy the old directory or entry from"
+          ])
+    <*> argument readPath ( mconcat
+          [ metavar "NEWPATH"
+          , help "The path to copy the directory or entry to"
+          ])
+    <*> switch ( mconcat
+          [ long "force"
+          , short 'f'
+          , help "If a directory or entry already exists under the specified path, delete it and insert this entry instead"
+          ])
+
+deleteOptions :: Parser DeleteOptions
+deleteOptions =
+  DeleteOptions
+    <$> argument readPath ( mconcat
+          [ metavar "PATH"
+          , help "The path to the entry or directory to delete"
+          ])
+    <*> switch ( mconcat
+          [ long "recursive"
+          , short 'r'
+          , help "If the specified path is a directory, remove it and its contents"
+          ])
+
+tagOptions :: Parser TagOptions
+tagOptions =
+  TagOptions
+    <$> argument readEntryPath ( mconcat
+          [ metavar "ENTRYPATH"
+          , help "The path to the entry to add or delete a tag from"
+          ])
+    <*> argument readEntryTag ( mconcat
+          [ metavar "TAGNAME"
+          , help "The name of the tag to add or delete"
+          ])
+    <*> switch ( mconcat
+          [ long "delete"
+          , short 'd'
+          , help "Remove this tag instead of setting it"
+          ])
+
+----------------------------------------------------------------------------
+-- Common
+----------------------------------------------------------------------------
+
+readPath :: ReadM Path
+readPath = do
+  input <- str
+  case mkPath input of
+    Right path -> pure path
+    Left err -> readerError $ unlines
+      [ "Invalid path: '" <> T.unpack input <> "'."
+      , T.unpack err
+      ]
```

Comment:
Why not [etherReader](https://hackage.haskell.org/package/optparse-applicative-0.17.0.0/docs/Options-Applicative.html#v:eitherReader)?

Reply:
Good question, I guess it would be more idiomatic
```hs
  eitherReader \input ->
    case mkPath (T.pack input) of
      Right path -> pure path
      Left err -> Left $ unlines
        [ "Invalid path: '" <> input <> "'."
        , T.unpack err
        ]
```

Could be shortened a bit more using `first` to map over the error:
```hs
  eitherReader \input ->
    mkPath (T.pack input) & first \err ->
      unlines
        [ "Invalid path: '" <> input <> "'."
        , T.unpack err
        ]
```